WikiWatershed/model-my-watershed
# Add Drainage Area Draw Tools

## Overview

Adds two new draw tools for Drainage Area, with dummy implementations. The first takes a point, and requests an area that drains to it. The second takes a polygon, drawn around streams, and requests and area that drains to those streams. The dummy implementation simply buffers the point or stream by 0.1 and returns the shape.

In upcoming changes, the source points / streams will be sent to the Drexel / ANS API, and we'll get the correct shapes from them.

The returned shape also has `drainage_area: true` property, which will be used to show special drainage area analysis.

Connects #3546 

### Demo

![2022-06-22 23 59 54](https://user-images.githubusercontent.com/1430060/175206989-f0a874b2-c715-4aa6-b3d5-41dd072d4a67.gif)

![2022-06-23 00 00 31](https://user-images.githubusercontent.com/1430060/175207017-e614ab79-d031-4c20-a409-5843e8e364a9.gif)

### Notes

In the current implementation the new draw endpoints are synchronous. However, we may need to make them Celery task based, as are all the other analyze and model endpoints, since the Drexel API may take a while to run.

There is also more validation that needs to be done, and error handling added, so that validation errors are communicated clearly to the user. For example, if in your stream selection shape, there are no streams, there is currently no graceful error handling for that.

## Testing Instructions

* Check out this branch and `./scripts/bundle.sh --debug`
* Go to http://localhost:8000/ and click Get Started
  - [x] Ensure you see a new Drainage Area set of draw tools
* Click the point-based drainage area draw tool
  - [x] Ensure it draws an oval around the point you clicked
  - [x] Ensure Analyze completes as usual
* Reset, and click the stream-based drainage area draw tool
  - [x] Ensure it turns on the streams layer for you to see where you're drawing
  - [x] Ensure after you draw the shape, it draws ovals around the streams in your selection
  - [x] Ensure Analyze completes as usual
